domain/relation/state/relation.go

```diff
@@ -1539,35 +1510,231 @@ AND    ae.application_uuid = $relationAndApplicationUUID.application_uuid
 	return relationSettings, nil
 }
 
-func (st *State) checkApplicationInRelation(
+// SetRelationApplicationSettings records settings for a specific application
+// relation combination.
+//
+// The following error types can be expected to be returned:
+//   - [relationerrors.ApplicationNotFoundForRelation] is returned if the
+//     application is not part of the relation.
+//   - [relationerrors.RelationNotFound] is returned if the relation UUID
+//     is not found.
+func (st *State) SetRelationApplicationSettings(
+	ctx context.Context,
+	relationUUID corerelation.UUID,
+	applicationID application.ID,
+	settings map[string]string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get the relation endpoint UUID.
+		endpointUUID, err := st.getRelationEndpointUUID(ctx, tx, relationUUID, applicationID)
+		if err != nil {
+			return errors.Errorf("getting relation endpoint uuid: %w", err)
+		}
+
+		if len(settings) == 0 {
+			// Return early, but after checking for not found errors, if there
+			// are no settings to update.
+			return nil
+		}
```

Comment:
todo: Move check to the service. It's part of the CommitHook facade call, so it'll be used very frequently. Twice per hook per unit.

If the relation doesn't exist at this point, there is something bigger going on that other code will catch.